[Mobile] Cover: Add Media Button when only colour is selected

antonis opened the pull request:
**Fixes** https://github.com/WordPress/gutenberg/issues/23877 (**parent** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2275)

## Related PRs:
This PR depends on https://github.com/WordPress/gutenberg/pull/23870

* `gutenberg-mobile` https://github.com/wordpress-mobile/gutenberg-mobile/pull/2513
* **Depends on** Create cover block with custom colour https://github.com/WordPress/gutenberg/pull/23870
* Create cover block with only colour from a default palette https://github.com/WordPress/gutenberg/pull/23833
* Clear media button in block settings https://github.com/WordPress/gutenberg/pull/23794

## Description
An image icon button has been added at the top-left of the component. When pressed the add media options are presented.

## How has this been tested?
* Add a cover block
* Press on a colour from the palette
* Verify that a new cover block with only colour is created
* Verify that the new icon is presented at the top-left of the component
* Tap on the icon and verify that the add media options are presented
* Add a media and verify that it is applied in the background

## Screenshots

|Screenshot|Functionality|
|---|---|
|![Simulator Screen Shot - iPhone 11 - 2020-07-10 at 18 51 16](https://user-images.githubusercontent.com/304044/87174203-826e8c80-c2df-11ea-8790-c8a623805be0.png)|![app](https://user-images.githubusercontent.com/304044/87912606-589f2d80-ca76-11ea-9e05-6bf0388b3f6e.gif)|

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

antonis committed:
Create cover block with only colour (from a default palette)

antonis committed:
Added custom colour swatch

antonis committed:
Custom Color picker action

antonis committed:
Hide the block settings button

antonis committed:
Add image button

antonis committed:
Use noop instead of empty arrow function

antonis committed:
Use default settings colors

antonis committed:
WIP: Changes Based on design feedback

antonis committed:
Merge branch 'master' into rnmobile/23831_Cover_Only_color

antonis committed:
Merge branch 'master' into rnmobile/23831_Cover_Only_color

antonis committed:
Merge branch 'rnmobile/23831_Cover_Only_color' into rnmobile/Cover_with_custom_color

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Show palette with both hardcoded and theme colors

antonis committed:
Customise media placeholder styles only for the cover component

antonis committed:
Merge branch 'rnmobile/23831_Cover_Only_color' into rnmobile/Cover_with_custom_color

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Pass only modified media-placeholder styles

antonis committed:
Pass only modified color-palette styles

antonis committed:
Merge branch 'rnmobile/23831_Cover_Only_color' into rnmobile/Cover_with_custom_color

# Conflicts:
#	packages/block-library/src/cover/style.native.scss
#	packages/components/src/color-palette/index.native.js

antonis committed:
Fixes lint issue

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Updated color palette height

antonis committed:
Updated color palette height

antonis committed:
Use only theme colors

antonis committed:
Merge branch 'rnmobile/23831_Cover_Only_color' into rnmobile/Cover_with_custom_color

antonis committed:
Merge branch 'master' into rnmobile/23831_Cover_Only_color

antonis committed:
Fixes Dark mode styling bug

antonis committed:
Merge branch 'rnmobile/23831_Cover_Only_color' into rnmobile/Cover_with_custom_color

antonis committed:
Use noop instead of empty arrow functions

antonis committed:
Simplified media-placeholder props

antonis committed:
Removed unnecessary style

antonis committed:
Revert "Removed unnecessary style"

This reverts commit 7d6219f96be23918d4e75822c6ac895c25651b29.

antonis committed:
Merge branch 'rnmobile/23831_tmp' into rnmobile/Cover_with_custom_color

antonis committed:
Merge branch 'rnmobile/23831_tmp' into rnmobile/Cover_with_custom_color

antonis committed:
Fixes style issue

antonis committed:
Added missing callbacks

antonis committed:
Added bottom label

antonis committed:
Removes unnecessary function

antonis committed:
Merge branch 'master' into rnmobile/Cover_with_custom_color

antonis committed:
Apply custom color picker color

antonis committed:
Removes unused import

antonis committed:
Lint issues and cleanup

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis commented:
Hello @iamthomasbishop 👋 ,
Please have a look at the media button addition and let me know wdyt.

iamthomasbishop commented:
Hi @antonis, can we use the same button style/sizing/spacing as is used on the edit button when a media item is attached (with a different icon, of course)? Example:

![image](https://user-images.githubusercontent.com/1200432/88080955-62dc2d00-cb45-11ea-939f-ff495bf33577.jpeg)

antonis committed:
Merge branch 'master' into rnmobile/Cover_with_custom_color

antonis committed:
Design fixes

antonis committed:
Renamed prop to declare usage more clearly

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Lint Issue

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Changed image button design

antonis committed:
Show image button only when component is selected

antonis committed:
Display correct media picker label

antonis commented:
Thank you for the feedback @iamthomasbishop :)

I updated the button design and changed the text label.
Since the MediaUpload bottom sheet component already had a text for this I kept it "Choose image or video"

"Light background|Dark background|
|---|---|
|![Simulator Screen Shot - iPhone 11 - 2020-07-23 at 10 10 29](https://user-images.githubusercontent.com/304044/88260465-191e4e80-cccd-11ea-8eb6-a66c3d3c1e50.png)|![Simulator Screen Shot - iPhone 11 - 2020-07-23 at 10 05 25](https://user-images.githubusercontent.com/304044/88260466-1ae81200-cccd-11ea-8444-dbcaf84ebac5.png)|

antonis committed:
PR review fixes

antonis committed:
Extra fixes to avoid inconsistent states

antonis committed:
Lint fix and comment

iamthomasbishop commented:
> I updated the button design and changed the text label.
> Since the MediaUpload bottom sheet component already had a text for this I kept it "Choose image or video"

Perfect. Thank you @antonis !

antonis committed:
Revert "Lint fix and comment"

This reverts commit 054a6b573cf264d3101135460745f3696fd02431.

antonis committed:
Revert "Extra fixes to avoid inconsistent states"

This reverts commit df25f6b0161b638b8bbf489ed8f97622ebaed8a4.

antonis committed:
Review changes

antonis committed:
Android hardware button handling

antonis committed:
Merge branch 'master' into rnmobile/Cover_with_custom_color

antonis committed:
Lint fix

antonis committed:
Merge branch 'rnmobile/Cover_with_custom_color' into rnmobile/CoverAddImageButton

antonis committed:
Syntactical fix

antonis committed:
Merge branch 'master' into rnmobile/CoverAddImageButton

antonis committed:
Merge fixes

mkevins reviewed (commented):
This is looking pretty good. I tested on Android, and the logic is working as expected. There is one issue with the button size / container size that makes tapping difficult. It seems that the tappable area is small, and offset from the button itself:

<img src="https://user-images.githubusercontent.com/8507675/88504849-3595f000-d019-11ea-9899-115a8403d241.gif" width="360">

I inspected the layout and it seems that the views are rendered outside their container (so they don't capture the event). E.g. the red rectangle is the outer view and the blue one is the child view:
<img width="772" alt="add-image-button-out-of-container" src="https://user-images.githubusercontent.com/8507675/88505184-1fd4fa80-d01a-11ea-99dd-29fb91b652e8.png">

It seems that aside from being offset, the size of the touchable is that of the inner view (inner view, e.g. icon, is the red rectangle):
<img width="772" alt="add-image-button-inner-view" src="https://user-images.githubusercontent.com/8507675/88505189-22cfeb00-d01a-11ea-8be4-1bf32bb777a3.png">

I tried locally changing the order (so `TouchableWithoutFeedback` wraps `View`), but the problem still persisted, so perhaps a change is needed elsewhere, either instead of or in addition to that.

antonis committed:
Merge branch 'master' into rnmobile/CoverAddImageButton

antonis committed:
Review fixes

antonis commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -330,6 +346,15 @@ const Cover = ( {
+			{ renderBackground( getMediaOptions ) }
+			{ isParentSelected &&
+				hasOnlyColorBackground &&
+				addMediaButton( openMediaOptionsRef.current ) }
```
Hello @mkevins 👋 
thank you for your feedback, I really appreciate it :)
You are right, the parameter can be accessed directly. I updated the PR changing this and fixing the tappable area.

mkevins commented on packages/block-library/src/cover/edit.native.js:
```diff
@@ -252,9 +252,12 @@ const Cover = ( {
 
-	const addMediaButton = ( open ) => (
+	const addMediaButton = () => (
 		<View style={ styles.selectImageContainer }>
```
We could remove the double `styles.selectImageContainer` by wrapping the view like this:

```patch
diff --git a/packages/block-library/src/cover/edit.native.js b/packages/block-library/src/cover/edit.native.js
index 14b413712a..d698791df7 100644
--- a/packages/block-library/src/cover/edit.native.js
+++ b/packages/block-library/src/cover/edit.native.js
@@ -253,11 +253,8 @@ const Cover = ( {
 	);
 
 	const addMediaButton = () => (
-		<View style={ styles.selectImageContainer }>
-			<TouchableWithoutFeedback
-				style={ styles.selectImageContainer }
-				onPress={ openMediaOptionsRef.current }
-			>
+		<TouchableWithoutFeedback onPress={ openMediaOptionsRef.current }>
+			<View style={ styles.selectImageContainer }>
 				<View style={ styles.selectImage }>
 					<Icon
 						size={ 16 }
@@ -265,8 +262,8 @@ const Cover = ( {
 						{ ...styles.selectImageIcon }
 					/>
 				</View>
-			</TouchableWithoutFeedback>
-		</View>
+			</View>
+		</TouchableWithoutFeedback>
 	);
 
 	const controls = (

```

antonis replied:
Thank you for the suggestion @mkevins! This looks much better. I updated the PR

mkevins reviewed (approved):
Looks good and it's working correctly now. Nice work @antonis ! :tada: 

I left one minor suggestion (as a patch, since the GitHub diff context couldn't capture it as a "one-click-suggestion"), not a blocker though.

antonis committed:
layout improvement
